mempool/cat/pool.go

```diff
@@ -0,0 +1,765 @@
+	sort.Slice(txs, func(i, j int) bool {
+		if txs[i].priority == txs[j].priority {
+			return txs[i].timestamp.Before(txs[j].timestamp)
+		}
+		return txs[i].priority > txs[j].priority // N.B. higher priorities first
+	})
+	return txs
+}
+
+// ReapMaxBytesMaxGas returns a slice of valid transactions that fit within the
+// size and gas constraints. The results are ordered by nonincreasing priority,
+// with ties broken by increasing order of arrival. Reaping transactions does
+// not remove them from the mempool
+//
+// If maxBytes < 0, no limit is set on the total size in bytes.
+// If maxGas < 0, no limit is set on the total gas cost.
+//
+// If the mempool is empty or has no transactions fitting within the given
+// constraints, the result will also be empty.
+func (txmp *TxPool) ReapMaxBytesMaxGas(maxBytes, maxGas int64) types.Txs {
+	var totalGas, totalBytes int64
+
+	var keep []types.Tx //nolint:prealloc
+	for _, w := range txmp.allEntriesSorted() {
+		// N.B. When computing byte size, we need to include the overhead for
+		// encoding as protobuf to send to the application.
+		totalGas += w.gasWanted
+		totalBytes += types.ComputeProtoSizeForTxs([]types.Tx{w.tx})
+		if (maxGas >= 0 && totalGas > maxGas) || (maxBytes >= 0 && totalBytes > maxBytes) {
+			break
+		}
+		keep = append(keep, w.tx)
+	}
+	return keep
+}
+
+// ReapMaxTxs returns up to max transactions from the mempool. The results are
+// ordered by nonincreasing priority with ties broken by increasing order of
+// arrival. Reaping transactions does not remove them from the mempool.
+//
+// If max < 0, all transactions in the mempool are reaped.
+//
+// The result may have fewer than max elements (possibly zero) if the mempool
+// does not have that many transactions available.
+func (txmp *TxPool) ReapMaxTxs(max int) types.Txs {
+	var keep []types.Tx //nolint:prealloc
+
+	for _, w := range txmp.allEntriesSorted() {
+		if max >= 0 && len(keep) >= max {
+			break
+		}
+		keep = append(keep, w.tx)
+	}
+	return keep
+}
+
+// Update removes all the given transactions from the mempool and the cache,
+// and updates the current block height. The blockTxs and deliverTxResponses
+// must have the same length with each response corresponding to the tx at the
+// same offset.
+//
+// If the configuration enables recheck, Update sends each remaining
+// transaction after removing blockTxs to the ABCI CheckTx method.  Any
+// transactions marked as invalid during recheck are also removed.
+//
+// The caller must hold an exclusive mempool lock (by calling txmp.Lock) before
+// calling Update.
+func (txmp *TxPool) Update(
+	blockHeight int64,
+	blockTxs types.Txs,
+	deliverTxResponses []*abci.ResponseDeliverTx,
+	newPreFn mempool.PreCheckFunc,
+	newPostFn mempool.PostCheckFunc,
+) error {
+	// Safety check: Transactions and responses must match in number.
+	if len(blockTxs) != len(deliverTxResponses) {
+		panic(fmt.Sprintf("mempool: got %d transactions but %d DeliverTx responses",
```

Comment:
what's the rationale for a `panic` here vs returning an error? If there is a mis-match, is it non-recoverable?

Reply:
It's a developer error. It should never happen because both this method and the caller of this method are within this repo i.e. the developer controls both sides